File: pkg/models/query.go
package models

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/grafana/grafana-plugin-sdk-go/backend"
)

// GlpiQueryType — debe coincidir exactamente con src/types.ts
type GlpiQueryType string

const (
	QueryTicketSummary      GlpiQueryType = "ticket_summary"
	QueryTicketTrend        GlpiQueryType = "ticket_trend"
	QueryTicketList         GlpiQueryType = "ticket_list"
	QueryTicketByCategory   GlpiQueryType = "ticket_by_category"
	QueryTicketByTechnician GlpiQueryType = "ticket_by_technician"
	QuerySLACompliance      GlpiQueryType = "sla_compliance"
	QuerySLABreaches        GlpiQueryType = "sla_breaches"
	QueryResolutionTime     GlpiQueryType = "resolution_time"
	QueryPriorityBreakdown  GlpiQueryType = "priority_breakdown"
	QueryAssetCount         GlpiQueryType = "asset_count"
	QueryAssetList          GlpiQueryType = "asset_list"
	QueryUserWorkload       GlpiQueryType = "user_workload"
	QueryEntitySummary      GlpiQueryType = "entity_summary"
	QuerySatisfaction       GlpiQueryType = "satisfaction"
)

// GlpiFilters — filtros comunes a todos los query types
type GlpiFilters struct {
	EntityID     *int  `json:"entity_id,omitempty"`
	Status       []int `json:"status,omitempty"`
	Priority     []int `json:"priority,omitempty"`
	CategoryID   *int  `json:"category_id,omitempty"`
	TechnicianID *int  `json:"technician_id,omitempty"`
	GroupID      *int  `json:"group_id,omitempty"`
	TicketType   *int  `json:"ticket_type,omitempty"`
}

type TicketTrendOptions struct {
	Interval string `json:"interval"` // hour|day|week|month
	GroupBy  string `json:"group_by"` // none|status|priority|category
}

type SlaOptions struct {
	SlaID          *int `json:"sla_id,omitempty"`
	ShowPercentage bool `json:"show_percentage"`
}

type ResolutionTimeOptions struct {
	Unit           string `json:"unit"`            // minutes|hours|days
	ShowPercentile bool   `json:"show_percentile"` // incluir p95
}

type AssetOptions struct {
	Itemtype string `json:"itemtype"` // Computer|Monitor|NetworkEquipment|...
}

// GlpiQueryModel — estructura completa del query desde el frontend
type GlpiQueryModel struct {
	QueryType   GlpiQueryType `json:"queryType"`
	Filters     GlpiFilters   `json:"filters"`
	GroupByEntity bool        `json:"groupByEntity"`
	MaxResults  int           `json:"maxResults"`

	TicketTrendOptions    *TicketTrendOptions    `json:"ticketTrendOptions,omitempty"`
	SlaOptions            *SlaOptions            `json:"slaOptions,omitempty"`
	ResolutionTimeOptions *ResolutionTimeOptions `json:"resolutionTimeOptions,omitempty"`
	AssetOptions          *AssetOptions          `json:"assetOptions,omitempty"`

	// Inyectados por el Go (no vienen del frontend)
	TimeFrom time.Time `json:"-"`
	TimeTo   time.Time `json:"-"`
}

func ParseQuery(query backend.DataQuery) (*GlpiQueryModel, error) {
	var qm GlpiQueryModel
	if err := json.Unmarshal(query.JSON, &qm); err != nil {
		return nil, fmt.Errorf("json unmarshal query: %w", err)
	}

	qm.TimeFrom = query.TimeRange.From
	qm.TimeTo = query.TimeRange.To

	if qm.MaxResults <= 0 {
		qm.MaxResults = 100
	}

	return &qm, nil
}

File: backend-py/routers/query.py
from fastapi import APIRouter
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from typing import Optional
from glpi_client import get_api, build_response
import handlers

router = APIRouter()

class Filters(BaseModel):
    entity_id: Optional[int] = None
    status: list[int] = []
    priority: list[int] = []
    category_id: Optional[int] = None
    technician_id: Optional[int] = None
    group_id: Optional[int] = None
    ticket_type: Optional[int] = None

class TrendOptions(BaseModel):
    interval: str = "day"
    group_by: str = "none"

class SlaOptions(BaseModel):
    sla_id: Optional[int] = None
    show_percentage: bool = True

class ResolutionOptions(BaseModel):
    unit: str = "hours"
    show_percentile: bool = False

class AssetOptions(BaseModel):
    itemtype: str = "Computer"

class QueryRequest(BaseModel):
    query_type: str
    filters: Filters = Filters()
    group_by_entity: bool = False
    max_results: int = 100
    time_from: str
    time_to: str

    glpi_url: str
    app_token: str
    user_token: str = ""
    username: str = ""
    password: str = ""
    verify_ssl: bool = True

    trend_options: Optional[TrendOptions] = None
    sla_options: Optional[SlaOptions] = None
    resolution_options: Optional[ResolutionOptions] = None
    asset_options: Optional[AssetOptions] = None

HANDLER_MAP = {
    "ticket_summary":       handlers.ticket_summary,
    "ticket_trend":         handlers.ticket_trend,
    "ticket_list":          handlers.ticket_list,
    "ticket_by_category":   handlers.ticket_by_category,
    "ticket_by_technician": handlers.ticket_by_technician,
    "sla_compliance":       handlers.sla_compliance,
    "sla_breaches":         handlers.sla_breaches,
    "resolution_time":      handlers.resolution_time,
    "priority_breakdown":   handlers.priority_breakdown,
    "asset_count":          handlers.asset_count,
    "asset_list":           handlers.asset_list,
    "user_workload":        handlers.user_workload,
    "entity_summary":       handlers.entity_summary,
    "satisfaction":         handlers.satisfaction,
}

@router.post("/query")
def query(req: QueryRequest):
    handler = HANDLER_MAP.get(req.query_type)
    if not handler:
        return JSONResponse(status_code=400, content={"error": f"Unknown query_type: {req.query_type}"})

    try:
        api = get_api(req)
        result = handler(api, req)
        api.logout()
        return result

    except Exception as e:
        return JSONResponse(status_code=500, content={"error": str(e)})
